runtime/src/main/java/org/corfudb/runtime/view/TableRegistry.java

```diff
@@ -178,17 +179,21 @@ void registerTable(@Nonnull String namespace,
                 log.debug("registerTable: new schema:"+tableDescriptors.getFileDescriptorsMap());
             }
         }
-        try {
-            this.runtime.getObjectsView().TXBuild().type(TransactionType.OPTIMISTIC).build().begin();
-            if (hasSchemaChanged) {
-                this.registryTable.put(tableNameKey,
-                        new CorfuRecord<>(tableDescriptors, metadataBuilder.build()));
-            } else {
-                this.registryTable.putIfAbsent(tableNameKey,
-                        new CorfuRecord<>(tableDescriptors, metadataBuilder.build()));
+        int numRetries = 3; // Since this is an internal transaction, retry a few times before giving up.
+        while (numRetries-- > 0) {
+            try {
+                this.runtime.getObjectsView().TXBuild().type(TransactionType.OPTIMISTIC).build().begin();
+                if (hasSchemaChanged) {
+                    this.registryTable.put(tableNameKey,
+                            new CorfuRecord<>(tableDescriptors, metadataBuilder.build()));
+                } else {
+                    this.registryTable.putIfAbsent(tableNameKey,
+                            new CorfuRecord<>(tableDescriptors, metadataBuilder.build()));
+                }
+                this.runtime.getObjectsView().TXEnd();
```

Comment:
Is it possible that before calling TXEnd(), while doing put, it will throw a transaction abort exception and the tx.end() is never called?

Reply:
No, we can't throw transaction abort exceptions until a transaction ends because only when a transaction ends will we have all the keys whose conflicts need resolution.

Reply:
Ok. How about other exceptions? If before call TXEnd(), the above code throws other exceptions but not transaction abort exception, the the TXEnd() will not be called, is it safe to start a transaction without a TXEnd() call? Also after 3 retries, if it still catch a transaction abort exception, it should throw it, right?

Reply:
Only TransactionAborted exception is caught and retried which is not thrown before TXEnd() is called as of this codebase.
If there are other exceptions then we will not catch or retry them, it will go up to caller as usual.

Reply:
Two issues:
1. The original code, always execute TXEnd() as it is in the finally. So with any exception, the TXEnd() will be executed. 
2. In the change, if transactionAborted happens three times after retry, the exception should be thrown again to the caller.

Reply:
1. Calling TXEnd() on an other exception that occurred before it would be as undefined/wrong behavior as calling TXEnd() in the finally block - (one might end up committing a transaction that was supposed to fail). I suppose it's better to abort the transaction in that case to prevent corruption. Let me make that change instead.
2. Sure, while this will almost never occur in real life because the transaction abort itself is rare, I suppose it should be cleaner to re-throw the exception.

Reply:
Addressed 1. and 2. thanks